src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs

```diff
@@ -30,6 +30,162 @@ internal void VerifyDiagnostics(string source, CSharpCompilationOptions options,
+        public void LocalFunctionDoesNotOverwriteInnerScopeFlags()
+        {
+            var source = @"
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+class C
+{
+    void M(List<Task> listOfTasks)
+    {
+        lock(listOfTasks)
+        {
+            async Task localFunc()
+            {
+                lock(null)
+                {
+                    await Task.Yield();
+                }
+            }
+
+            listOfTasks.Add(localFunc());
+        }
+
+        try
+        {
+            IEnumerable<int> localFunc()
+            {
+                try
+                {
+                    yield return 1;
+                }
+                catch (Exception){}
+            }
+
+            localFunc();
+        }
+        catch (Exception)
+        {
+            IEnumerable<int> localFunc()
+            {
+                try{}
+                catch (Exception)
+                {
+                    yield return 1;
+                }
+
+                try{}
+                catch (Exception)
+                {
+                    throw;
+                }
+            }
+
+            localFunc();
+        }
+        finally
+        {
+            IEnumerable<int> localFunc()
+            {
+                try{}
+                finally
+                {
+                    yield return 1;
+                }
+            }
+
+            localFunc();
+        }
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib46(source);
+            comp.VerifyDiagnostics(
+                // (16,21): error CS1996: Cannot await in the body of a lock statement
+                //                     await Task.Yield();
+                Diagnostic(ErrorCode.ERR_BadAwaitInLock, "await Task.Yield()").WithLocation(16, 21),
+                // (29,21): error CS1626: Cannot yield a value in the body of a try block with a catch clause
+                //                     yield return 1;
+                Diagnostic(ErrorCode.ERR_BadYieldInTryOfCatch, "yield").WithLocation(29, 21),
+                // (43,21): error CS1631: Cannot yield a value in the body of a catch clause
+                //                     yield return 1;
+                Diagnostic(ErrorCode.ERR_BadYieldInCatch, "yield").WithLocation(43, 21),
+                // (62,21): error CS1625: Cannot yield in the body of a finally clause
+                //                     yield return 1;
+                Diagnostic(ErrorCode.ERR_BadYieldInFinally, "yield").WithLocation(62, 21));
+        }
+
+        [ConditionalFact(typeof(DesktopOnly))]
+        public void LocalFunctionResetsScopeFlags()
+        {
+            var source = @"
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+class C
+{
+    void M(List<Task> listOfTasks)
+    {
+        lock(listOfTasks)
+        {
+            async Task localFunc()
+            {
+                await Task.Yield();
```

Comment:
>await Task.Yield(); [](start = 16, length = 19)

I think Vlad pointed this out earlier. You should split the test into the portion that works and the portion that produces a diagnostic. Then the portion that works should be checked for execution (`CompileAndVerify(..., expectedOutput: "")`).

Reply:
Done.